### ISSUE TITLE
Inconsistent gradients through torch.symeig

### ISSUE THREAD

**Issue by calincru** [pytorch/pytorch#22807]
## 🐛 Bug

The gradients through `torch.symeig` are not consistent with the recent change that (rightfully) made the gradients through Cholesky decomposition symmetric (see https://github.com/pytorch/pytorch/issues/18825).  The same change should be applied for symmetric eigenvalue decomposition.

See also my comment https://github.com/pytorch/pytorch/issues/18825#issuecomment-510958045. In general I believe that it makes no sense for an implementation detail (i.e., which part of the symmetric matrix the algorithms looks at) to reflect into the returned gradients.

## To Reproduce

Here's a simple snippet that computes the logarithm of the determinant of a positive definite matrix in two different ways, yielding gradients that are different only in that one (which is correct) is the symmetrized version of the other (the upper/lower diagonal one).

```python
import torch

mat = torch.randn(4, 4, dtype=torch.float64)
mat = (mat @ mat.transpose(-1, -2)).div_(2).add_(torch.eye(4, dtype=torch.float64))
mat = mat.detach().clone().requires_grad_(True)
mat_clone = mat.detach().clone().requires_grad_(True)

# Way 1
chol_mat = mat.cholesky()
logdet1 = 2 * chol_mat.diagonal().log().sum()

# Way 2
w, _ = mat_clone.symeig(eigenvectors=True)
logdet2 = w.log().sum()

print('Are these both log(det(A))?', bool(logdet1 - logdet2 < 1e-8))  # True

logdet1.backward()
logdet2.backward()

inv_mat = mat.inverse()  # the analytical solution
print('Does Way 1 yield A^{-1}?', bool(torch.norm(mat.grad - inv_mat) < 1e-8))  # True
print('Does Way 2 yield A^{-1}?', bool(torch.norm(mat_clone.grad - inv_mat) < 1e-8))  # False
```

## Expected behavior

The returned gradients should be symmetric matrices so as to match the correct analytical derivative.

CC @vishwakftw

**Comment by calincru** [comment]
BTW, if there are legit use cases that expect upper or lower diagonal gradients, perhaps a better solution (both for `torch.symeig` and `torch.cholesky`) would be to make the `upper` argument ternary. That is, if it's not provided (e.g., `None`), return the canonical symmetric version. Otherwise, return the upper/lower gradients according to its boolean value.

It might be worthwhile for you to check if this makes sense in other scenarios too.

**Comment by vishwakftw** [comment]
If you require triangular gradients, you can always use a triu / tril op to get an upper or a lower triangular gradient. Making the upper argument ternary is not really viable, but I’m not sure.

**Comment by fmassa** [comment]
@vishwakftw what is the action that you think should be done here? Should we modify the implementation of the derivative of `symeig`?

**Comment by vishwakftw** [comment]
@fmassa sorry I couldn’t get this to earlier. I think we will have to modify the implementation of the gradients to make them symmetric. I can send in a PR with this fix tomorrow, if that’s fine.

**Comment by vishwakftw** [comment]
Closed by #23018
